File: lib/java/src/main/java/io/github/telepact/internal/schema/BuildDocumentLocatorFromYamlAst.java
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package io.github.telepact.internal.schema;

import java.io.StringReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.fasterxml.jackson.databind.ObjectMapper;

import org.yaml.snakeyaml.Yaml;
import org.yaml.snakeyaml.error.Mark;
import org.yaml.snakeyaml.nodes.MappingNode;
import org.yaml.snakeyaml.nodes.Node;
import org.yaml.snakeyaml.nodes.NodeTuple;
import org.yaml.snakeyaml.nodes.ScalarNode;
import org.yaml.snakeyaml.nodes.SequenceNode;

import io.github.telepact.internal.schema.DocumentLocators.DocumentLocator;

public class BuildDocumentLocatorFromYamlAst {
    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

    public static DocumentLocator createDocumentLocatorFromYamlText(String text) throws Exception {
        final var root = new Yaml().compose(new StringReader(text));
        if (root == null) {
            throw new IllegalArgumentException("YAML document cannot be empty");
        }

        validateNoDuplicateKeys(root);
        final var locations = new HashMap<String, Map<String, Object>>();
        locations.put(serializePath(List.of()), coordinatesFromMark(root.getStartMark()));
        buildLocationsForNode(root, List.of(), locations);

        return path -> new HashMap<>(locations.getOrDefault(serializePath(path), defaultDocumentCoordinates()));
    }

    private static void buildLocationsForNode(
            Node node,
            List<Object> path,
            Map<String, Map<String, Object>> locations) throws Exception {
        if (node instanceof SequenceNode sequenceNode) {
            final var values = sequenceNode.getValue();
            for (int index = 0; index < values.size(); index += 1) {
                final var child = values.get(index);
                final var childPath = new ArrayList<>(path);
                childPath.add(index);
                locations.put(serializePath(childPath), coordinatesFromMark(child.getStartMark()));
                buildLocationsForNode(child, childPath, locations);
            }
            return;
        }

        if (node instanceof MappingNode mappingNode) {
            for (final NodeTuple tuple : mappingNode.getValue()) {
                if (!(tuple.getKeyNode() instanceof ScalarNode keyNode)) {
                    continue;
                }

                final var keyPath = new ArrayList<>(path);
                keyPath.add(keyNode.getValue());
                locations.put(serializePath(keyPath), coordinatesFromMark(keyNode.getStartMark()));
                buildLocationsForNode(tuple.getValueNode(), keyPath, locations);
            }
        }
    }

    private static void validateNoDuplicateKeys(Node node) {
        if (node instanceof SequenceNode sequenceNode) {
            for (final var child : sequenceNode.getValue()) {
                validateNoDuplicateKeys(child);
            }
            return;
        }

        if (node instanceof MappingNode mappingNode) {
            final var seenKeys = new java.util.HashSet<String>();
            for (final NodeTuple tuple : mappingNode.getValue()) {
                final var key = tuple.getKeyNode() instanceof ScalarNode scalarNode
                        ? scalarNode.getValue()
                        : String.valueOf(tuple.getKeyNode());
                if (!seenKeys.add(key)) {
                    throw new IllegalArgumentException("Duplicate YAML key");
                }
                validateNoDuplicateKeys(tuple.getValueNode());
            }
        }
    }

    private static Map<String, Object> coordinatesFromMark(Mark mark) {
        if (mark == null) {
            return defaultDocumentCoordinates();
        }
        return coordinates(mark.getLine() + 1, mark.getColumn() + 1);
    }

    private static Map<String, Object> coordinates(int row, int col) {
        return Map.of("row", row, "col", col);
    }

    private static Map<String, Object> defaultDocumentCoordinates() {
        return coordinates(1, 1);
    }

    private static String serializePath(List<Object> path) {
        try {
            return OBJECT_MAPPER.writeValueAsString(path);
        } catch (Exception e) {
            return "[]";
        }
    }
}

File: lib/py/telepact/internal/schema/BuildDocumentLocatorFromYamlAst.py
#|
#|  Copyright The Telepact Authors
#|
#|  Licensed under the Apache License, Version 2.0 (the "License");
#|  you may not use this file except in compliance with the License.
#|  You may obtain a copy of the License at
#|
#|  https://www.apache.org/licenses/LICENSE-2.0
#|
#|  Unless required by applicable law or agreed to in writing, software
#|  distributed under the License is distributed on an "AS IS" BASIS,
#|  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#|  See the License for the specific language governing permissions and
#|  limitations under the License.
#|

import json
from typing import Callable

import yaml


def _serialize_path(path: list[object]) -> str:
    return json.dumps(path, separators=(',', ':'))


def _coordinates_from_mark(mark: yaml.Mark | None) -> dict[str, object]:
    if mark is None:
        return {'row': 1, 'col': 1}
    return {'row': mark.line + 1, 'col': mark.column + 1}


def _build_locations_for_node(
    node: yaml.nodes.Node | None,
    path: list[object],
    locations: dict[str, dict[str, object]],
) -> None:
    if isinstance(node, yaml.nodes.SequenceNode):
        for index, item in enumerate(node.value):
            item_path = path + [index]
            locations[_serialize_path(item_path)] = _coordinates_from_mark(item.start_mark)
            _build_locations_for_node(item, item_path, locations)
        return

    if isinstance(node, yaml.nodes.MappingNode):
        for key_node, value_node in node.value:
            key = getattr(key_node, 'value', None)
            if not isinstance(key, str):
                continue

            key_path = path + [key]
            locations[_serialize_path(key_path)] = _coordinates_from_mark(key_node.start_mark)
            _build_locations_for_node(value_node, key_path, locations)


def _validate_no_duplicate_keys(node: yaml.nodes.Node | None) -> None:
    if isinstance(node, yaml.nodes.SequenceNode):
        for item in node.value:
            _validate_no_duplicate_keys(item)
        return

    if isinstance(node, yaml.nodes.MappingNode):
        seen_keys: set[str] = set()
        for key_node, value_node in node.value:
            key = str(getattr(key_node, 'value', ''))
            if key in seen_keys:
                raise ValueError('Duplicate YAML key')
            seen_keys.add(key)
            _validate_no_duplicate_keys(value_node)


def create_document_locator_from_yaml_text(text: str) -> Callable[[list[object]], dict[str, object]]:
    node = yaml.compose(text)
    if node is None:
        raise ValueError('YAML document cannot be empty')

    _validate_no_duplicate_keys(node)
    locations: dict[str, dict[str, object]] = {
        _serialize_path([]): _coordinates_from_mark(node.start_mark),
    }
    _build_locations_for_node(node, [], locations)

    return lambda path: locations.get(_serialize_path(path), {'row': 1, 'col': 1})
